### PR TITLE
Locust

### PR DESCRIPTION
**Pull request by willronchetti**
- Provides a simple, configurable load testing interface for use on FF

### CODE REVIEW

File: src/encoded/locust/main.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument('config', help='path to config.json')
+    parser.add_argument('key', help='path to <env>.json')
+    parser.add_argument('--time', help='time to run test for, default 1m. Format: 10s, 5m, 1h, 1h30m etc.')
+    parser.add_argument('--clients', help='number of clients, default 10')
+    parser.add_argument('--rate', help='number of clients to hatch per second, default 10')
+    parser.add_argument('--lower', help='lower bound on time to wait between requests, default 1')
+    parser.add_argument('--upper', help='upper bound on time to wait between requests, default 2')
+
```
Comment (j1z0) [inline src/encoded/locust/main.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/encoded/locust/main.py  6
- src/encoded/locust/ff_locust.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435&bid=15687383)

File: src/encoded/locust/ff_locust.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+class Locust(HttpLocust):
+    """ test_set defines what the tasks are to be executed by the locust
+        host is where to test
+        wait_time is how long to wait between requests (5-15 seconds)
+
+        For cmd line args, see main.py
+    """
+    task_set = FFUserBehavior
+    host = FFUserBehavior.CONFIG['envs'][os.environ['LOCUST_ENV']]
+    wait_time = between(int(os.environ['LOCUST_LOWER_BOUND']), int(os.environ['LOCUST_UPPER_BOUND']))
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import random
+from requests.auth import HTTPBasicAuth
+from locust import HttpLocust, TaskSet, task, between
+
+# IMPORTANT:
+# To best make use of this you'll want to pass in some authentication
+# To do this, create the file '<env>.json'. Navigate to the portal and
+# create an access key to be used via basic auth. Set 'username' and 'password'
+# in <env>.json to your ID and secret
+# config.json is provided. Use accordingly if you want to change routes, see
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        self.client.get(url, auth=HTTPBasicAuth(self.USERNAME, self.PASSWORD))
+
+    @task(1)
+    def random_task(self):
+        """ get a random task from config """
+        url = random.choice(self.ROUTES)
+        self._get(url)
+
+
+class Locust(HttpLocust):
```
Comment (netsettler) [inline src/encoded/locust/ff_locust.py]: Should this perhaps be called FFLocust? Or even just MyLocust or something that suggests it's doing something more specific than HttpLocust, not something more general, right? The naming you picked goes more general and is the kind of name I'd give if I were trying to first repair a class before then creating an application-specific use.

File: src/encoded/locust/ff_locust.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+# IMPORTANT:
+# To best make use of this you'll want to pass in some authentication
+# To do this, create the file '<env>.json'. Navigate to the portal and
+# create an access key to be used via basic auth. Set 'username' and 'password'
+# in <env>.json to your ID and secret
+# config.json is provided. Use accordingly if you want to change routes, see 
+# locust.rst in docs
+
+
+def load_auth(loc):
```
Comment (netsettler) [inline src/encoded/locust/ff_locust.py]: If this is going to be a general purpose load_auth, I'd rather it be in some separate utility package. If it's only used here, maybe it should be a static method on FFUserBehavior?

File: src/encoded/locust/main.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        description='Locust Load Testing',
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument('config', help='path to config.json')
+    parser.add_argument('key', help='path to <env>.json')
+    parser.add_argument('--time', help='time to run test for, default 1m. Format: 10s, 5m, 1h, 1h30m etc.')
+    parser.add_argument('--clients', help='number of clients, default 10')
+    parser.add_argument('--rate', help='number of clients to hatch per second, default 10')
+    parser.add_argument('--lower', help='lower bound on time to wait between requests, default 1')
+    parser.add_argument('--upper', help='upper bound on time to wait between requests, default 2')
```
Comment (netsettler) [inline src/encoded/locust/main.py]: Does it not work to do the following?
```
    parser.add_argument('--lower', help='lower bound on time to wait between requests, default 1', default=1, type=int)
    parser.add_argument('--upper', help='upper bound on time to wait between requests, default 2', default=2, type=int)
```

File: src/encoded/locust/main.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  
+    args = parser.parse_args()
+
+    # set env variables (this is the only way to do this) and configure args
+    env_copy = os.environ.copy()
+    env_copy['LOCUST_CONFIG'] = args.config
+    env_copy['LOCUST_KEY'] = args.key
+    env_copy['LOCUST_ENV'] = args.key.split('.')[0]  # data.json -> data
+    env_copy['LOCUST_LOWER_BOUND'] = args.lower or '1'
+    env_copy['LOCUST_UPPER_BOUND'] = args.upper or '2'
```
Comment (netsettler) [inline src/encoded/locust/main.py]: If you did the defaulting I suggested in in `.add_argument` above, would these `or <n>` things be needed?

File: src/encoded/locust/main.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    clients = args.clients or '10'
+    rate = args.rate or '10'
+    time = args.time or '1m'
```
Comment (netsettler) [inline src/encoded/locust/main.py]: Same issues about defaulting, though I didn't do the suggestion.

File: src/encoded/locust/main.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    Entry point for locust load testing. Run with -h to see arguments
+    """
+    parser = argparse.ArgumentParser(
+        description='Locust Load Testing',
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument('config', help='path to config.json')
+    parser.add_argument('key', help='path to <env>.json')
+    parser.add_argument('--time', help='time to run test for, default 1m. Format: 10s, 5m, 1h, 1h30m etc.')
+    parser.add_argument('--clients', help='number of clients, default 10')
```
Comment (netsettler) [inline src/encoded/locust/main.py]: I have a slight preference for calling the arguments `-nclients` or some such thing (maybe even with a short name of `-n`). The reason is that I vaguely expect an argument named `--clients` to have me list what the clients are, and so to feel like I'll have to go look some syntax up. Something that more strongly suggests a number feels better. I admit that I usually also prefer ordinary English words and can't defend `nclients` under that theory, but somehow it feels good to me. I'll leave all this to your best judgment. I just wanted to mention the issue.

(And see other comments for suggestions about defaults.)

File: src/encoded/locust/main.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    parser.add_argument('--lower', help='lower bound on time to wait between requests, default 1')
+    parser.add_argument('--upper', help='upper bound on time to wait between requests, default 2')
+  
+    args = parser.parse_args()
+
+    # set env variables (this is the only way to do this) and configure args
+    env_copy = os.environ.copy()
+    env_copy['LOCUST_CONFIG'] = args.config
+    env_copy['LOCUST_KEY'] = args.key
+    env_copy['LOCUST_ENV'] = args.key.split('.')[0]  # data.json -> data
```
Comment (netsettler) [inline src/encoded/locust/main.py]: Should this be `os.path.splitext(os.path.basename(args.key))[0]`? I find it annoying there's not a better way with `os.path` to get at the part before the extinction, but so it goes. The thing you're doing, but the main difference is that by using the file operations you make yourself immune to someone using a file specifier, and you alert the human reader of your code to the fact that the `args.key` is a pathname, which I _think_ it is. :)

File: src/encoded/locust/main.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    env_copy['LOCUST_LOWER_BOUND'] = args.lower or '1'
+    env_copy['LOCUST_UPPER_BOUND'] = args.upper or '2'
+    clients = args.clients or '10'
+    rate = args.rate or '10'
+    time = args.time or '1m'
+
+    # invoke locust
+    subprocess.call(['locust', '-f', 'ff_locust.py', '--no-web', '-c', clients, '-r', rate,
+                      '--run-time', time, '--print-stats'], env=env_copy)
```
Comment (netsettler) [inline src/encoded/locust/main.py]: I think I would rename `env_copy` to `locust_env`. It is indeed important to copy it before assigning it, but by the time of this call, it is no longer a copy, it's a modified copy, and so a logically distinct environment. The relevant part now that it's initialized is that it is what to use for `locust`.

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- src/encoded/locust/main.py  1
- src/encoded/locust/ff_locust.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435&bid=15687383)

File: src/encoded/locust/ff_locust.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+# To best make use of this you'll want to pass in some authentication
+# To do this, create the file '<env>.json'. Navigate to the portal and
+# create an access key to be used via basic auth. Set 'username' and 'password'
+# in <env>.json to your ID and secret
+# config.json is provided. Use accordingly if you want to change routes, see 
+# locust.rst in docs
+
+
+def load_locust_auth(loc):
+    """
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/main.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument('config', help='path to config.json')
+    parser.add_argument('key', help='path to <env>.json')
+    parser.add_argument('--time', help='time to run test for, default 1m. Format: 10s, 5m, 1h, 1h30m etc.', type=str, default='1m')
+    parser.add_argument('--nclients', help='number of clients, default 10', type=int, default=10)
+    parser.add_argument('--rate', help='number of clients to hatch per second, default 10', type=int, default=10)
+    parser.add_argument('--lower', help='lower bound on time to wait between requests, default 1', type=int, default=1)
+    parser.add_argument('--upper', help='upper bound on time to wait between requests, default 2', type=int, default=2)
+
```
Comment (j1z0) [inline src/encoded/locust/main.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import random
+from requests.auth import HTTPBasicAuth
+from locust import HttpLocust, TaskSet, task, between
+
+# IMPORTANT:
+# To best make use of this you'll want to pass in some authentication
+# To do this, create the file '<env>.json'. Navigate to the portal and
+# create an access key to be used via basic auth. Set 'username' and 'password'
+# in <env>.json to your ID and secret
+# config.json is provided. Use accordingly if you want to change routes, see
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+class FFLocust(HttpLocust):
+    """ test_set defines what the tasks are to be executed by the locust
+        host is where to test
+        wait_time is how long to wait between requests (5-15 seconds)
+
+        For cmd line args, see main.py
+    """
+    task_set = FFUserBehavior
+    host = FFUserBehavior.CONFIG['envs'][os.environ['LOCUST_ENV']]
+    wait_time = between(int(os.environ['LOCUST_LOWER_BOUND']), int(os.environ['LOCUST_UPPER_BOUND']))
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+# To do this, create the file '<env>.json'. Navigate to the portal and
+# create an access key to be used via basic auth. Set 'username' and 'password'
+# in <env>.json to your ID and secret
+# config.json is provided. Use accordingly if you want to change routes, see 
+# locust.rst in docs
+
+
+def load_locust_auth(loc):
+    """ 
+    Loads access key from given location
```
Comment (j1z0) [inline src/encoded/locust/ff_locust.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692435)

File: src/encoded/locust/ff_locust.py (hunk 4 above)
Comment (willronchetti) [reply src/encoded/locust/ff_locust.py]: Correct. Done.

File: src/encoded/locust/ff_locust.py (hunk 5 above)
Comment (willronchetti) [reply src/encoded/locust/ff_locust.py]: I changed this to ``load_locust_auth`` to differentiate. I left a comment about this but FFUserBehavior is implicitly a static class so I'm not sure it makes sense to pull that function in

File: src/encoded/locust/main.py (hunk 9 above)
Comment (willronchetti) [reply src/encoded/locust/main.py]: I like ``nclients`` as well and use the convention oftentimes myself. Changed.

File: src/encoded/locust/main.py (hunk 11 above)
Comment (willronchetti) [reply src/encoded/locust/main.py]: Great point. Changed

**Pull request merged**